client/asset/btc/spv.go

```diff
@@ -0,0 +1,1601 @@
+	if blockHeight < 0 {
+		var err error
+		blockHeight, err = w.cl.GetBlockHeight(blockHash)
+		if err != nil {
+			w.log.Errorf("Error getting block height for hash %s", blockHash)
+			return false
+		}
+	}
+	checkHash, err := w.cl.GetBlockHash(int64(blockHeight))
+	if err != nil {
+		w.log.Errorf("Error retriving block hash for height %d", blockHeight)
+		return false
+	}
+
+	return *checkHash == *blockHash
+}
+
+// mainchainBlockForStoredTx gets the block hash and height for the transaction
+// IFF an entry has been stored in the blockTxs index.
+func (w *spvWallet) mainchainBlockForStoredTx(txHash *chainhash.Hash) (*chainhash.Hash, int32) {
+	// Check that the block is still mainchain.
+	blockHash, blockHeight, err := w.blockForStoredTx(txHash)
+	if err != nil {
+		w.log.Errorf("Error retrieving mainchain block height for hash %s", blockHash)
+		return nil, 0
+	}
+	if blockHash == nil {
+		return nil, 0
+	}
+	if !w.blockIsMainchain(blockHash, blockHeight) {
+		return nil, 0
+	}
+	return blockHash, blockHeight
+}
+
+// findBlockForTime locates a good start block so that a search beginning at the
+// returned block has a very low likelyhood of missing any blocks that have time
+// > matchTime. This is done by performing a binary search (sort.Search) to find
+// a block with a block time MAX_FUTURE_BLOCK_TIME before matchTime. To ensure
+// we also accommodate the median-block time rule and aren't missing anything
+// due to out of sequence block times we use an unsophisticated algorithm of
+// choosing the first block in an 11 block window with no times >= matchTime.
+func (w *spvWallet) findBlockForTime(matchTime time.Time) (*chainhash.Hash, int32, error) {
+	offsetTime := matchTime.Add(-maxBlockTimeOffset)
+
+	bestHeight, err := w.getBestBlockHeight()
+	if err != nil {
+		return nil, 0, fmt.Errorf("getBestBlockHeight error: %v", err)
+	}
+
+	getBlockTimeForHeight := func(height int32) (*chainhash.Hash, time.Time, error) {
+		hash, err := w.cl.GetBlockHash(int64(height))
+		if err != nil {
+			return nil, time.Time{}, err
+		}
+		header, err := w.cl.GetBlockHeader(hash)
+		if err != nil {
+			return nil, time.Time{}, err
+		}
+		return hash, header.Timestamp, nil
+	}
+
+	iHeight := sort.Search(int(bestHeight), func(h int) bool {
+		var iTime time.Time
+		_, iTime, err = getBlockTimeForHeight(int32(h))
+		if err != nil {
+			return true
+		}
+		return iTime.After(offsetTime)
+	})
+	if err != nil {
+		return nil, 0, fmt.Errorf("binary search error finding best block for time %q: %w", matchTime, err)
+	}
+
+	// We're actually breaking an assumption of sort.Search here because block
+	// times aren't always monotonically increasing. This won't matter though
+	// as long as there are not > blockTimeSearchBuffer blocks with inverted
+	// time order.
+	var count int
+	var iHash *chainhash.Hash
+	var iTime time.Time
+	for iHeight > 0 {
+		iHash, iTime, err = getBlockTimeForHeight(int32(iHeight))
```

Comment:
I think different binary searches are likely to call `getBlockTimeForHeight` for the same heights.  I might instrument these closures with a bunch of prints to get a bit more intuition on the process, but we may benefit from caching block hashes and timestamps just within `findBlockForTime` to we don't have repeated RPCs, not that they will be costly.